### PR TITLE
[PIN-3257] Negate the possibility to creates two purposes with the same name in the same agreement

### PR DESCRIPTION
**Pull request by rGregnanin**
Added the functionality that prevents to create two purposes with the same name in one agreement

### CODE REVIEW

File: src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala [hunk 1]
```diff
@@ -111,8 +113,9 @@ object ReadModelQueries {
     val statesFilter       = mapToVarArgs(statesPartialFilter ++ archivedStatePartialFilter)(Filters.or)
     val eServicesIdsFilter = mapToVarArgs(eServicesIds.map(Filters.eq("data.eserviceId", _)))(Filters.or)
     val consumersIdsFilter = mapToVarArgs(consumersIds.map(Filters.eq("data.consumerId", _)))(Filters.or)
-    val nameFilter         = name.map(Filters.regex("data.title", _, "i"))
-
+    val nameFilter         =
+      if (exactMatchOnTitle) name.map(n => Filters.regex("data.name", s"^$n$$", "i"))
+      else name.map(Filters.regex("data.name", _, "i"))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala]:
```suggestion
      if (exactMatchOnTitle) name.map(n => Filters.regex("data.title", s"^$n$$", "i"))
      else name.map(Filters.regex("data.title", _, "i"))
```

File: src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala [hunk 2]
```diff
@@ -26,9 +26,10 @@ object ReadModelQueries {
     states: List[PurposeVersionState],
     excludeDraft: Boolean,
     offset: Int,
-    limit: Int
+    limit: Int,
+    exactMatchOnTitle: Boolean = false
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala:30]: this name should match the parameter it is related to, e.g.
- `exactMatchOnTitle` if the other param is `title`
- `exactMatchOnName` if the other param is `name`

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 3]
```diff
@@ -108,9 +108,27 @@ final case class PurposeApiServiceImpl(
...
+          List(agreement.eserviceId.toString),
+          List(agreement.consumerId.toString),
+          List(agreement.producerId.toString),
+          List.empty,
+          false,
+          0,
+          1,
+          exactMatchOnTitle = true
+        )(readModel)
+        .map(_.results.headOption.map(_.title))
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: No more required
```suggestion
        .map(_.results.headOption)
```

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 4]
```diff
@@ -108,9 +108,27 @@ final case class PurposeApiServiceImpl(
...
+
+      // _ <- if (maybePurpose.contains(seed.title)) Future.failed(DuplicatedPurposeName(seed.title)) else Future.unit
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: 🧹

File: src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala [hunk 5]
```diff
@@ -108,9 +108,27 @@ final case class PurposeApiServiceImpl(
...
+        .listPurposes(
+          seed.consumerId,
+          seed.title.some,
+          List(agreement.eserviceId.toString),
+          List(agreement.consumerId.toString),
+          List(agreement.producerId.toString),
+          List.empty,
+          false,
+          0,
+          1,
```
Comment (galales) [inline src/main/scala/it/pagopa/interop/purposeprocess/api/impl/PurposeApiServiceImpl.scala]: Just for readability
```suggestion
          states = List.empty,
          excludeDraft = false,
          offset = 0,
          limit = 1,
```

File: src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala (hunk 2 above)
Comment (TonioGela) [reply src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala:30]: +1

**Review by TonioGela — APPROVED** [review]
LGTM

File: src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala (hunk 2 above)
Comment (galales) [reply src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala:30]: this is not solved yet

Comment (beetlecrunch) [reply src/main/scala/it/pagopa/interop/purposeprocess/common/readmodel/ReadModelQueries.scala:30]: +1

**Review by beetlecrunch — CHANGES_REQUESTED** [review]
[This](https://github.com/pagopa/interop-be-purpose-process/pull/158#discussion_r1196046903) must be solved.

**Review by TonioGela — APPROVED** [review]

**Review by beetlecrunch — APPROVED** [review]

**Pull request merged**